crypto/machine.go

```diff
@@ -180,15 +190,192 @@ func (mach *OlmMachine) HandleToDeviceEvent(evt *event.Event) {
+			if content.RequestingDeviceID == mach.Client.DeviceID {
+				mach.Log.Debug("Ignoring key request from the same device (%v)", content.RequestingDeviceID)
+				return
+			}
+			mach.Log.Debug("Received key request from %v for session %v", content.RequestingDeviceID, content.Body.SessionID)
+
+			igs, err := mach.CryptoStore.GetGroupSession(content.Body.RoomID, content.Body.SenderKey, content.Body.SessionID)
+			if err != nil {
+				mach.Log.Error("Error retrieving IGS to forward key: %v", err)
+				return
+			}
+			if igs == nil {
+				mach.Log.Warn("Can't find the requested session to forward: %v", content.Body.SessionID)
+				return
+			}
+
+			// export session key at earliest index
+			exportedKey, err := igs.Internal.Export(igs.Internal.FirstKnownIndex())
+			if err != nil {
+				mach.Log.Error("Error exporting key for session %v: %v", igs.ID(), err)
+				return
+			}
+
+			forwardedRoomKey := event.Content{
+				Parsed: &event.ForwardedRoomKeyEventContent{
+					RoomKeyEventContent: event.RoomKeyEventContent{
+						Algorithm:  id.AlgorithmMegolmV1,
+						RoomID:     igs.RoomID,
+						SessionID:  igs.ID(),
+						SessionKey: exportedKey,
+					},
+					SenderKey:          content.Body.SenderKey,
+					ForwardingKeyChain: igs.ForwardingChains,
+					SenderClaimedKey:   igs.SigningKey,
+				},
+			}
+
+			if err := mach.SendEncryptedToDevice(evt.Sender, content.RequestingDeviceID, forwardedRoomKey); err != nil {
+				mach.Log.Error("Failed to send encrypted forwarded key: %v", err)
+			}
+
+			mach.Log.Debug("Sent encrypted forwarded key to device %v for session %v", content.RequestingDeviceID, igs.ID())
+		}
+	// TODO add m.verification.start case
 	default:
 		deviceID, _ := evt.Content.Raw["device_id"].(string)
 		mach.Log.Trace("Unhandled to-device event of type %s from %s/%s", evt.Type.Type, evt.Sender, deviceID)
 	}
 }
 
+// SendEncryptedToDevice sends an Olm-encrypted event to the given user device.
+func (mach *OlmMachine) SendEncryptedToDevice(userID id.UserID, deviceID id.DeviceID, content event.Content) error {
+	// get device identity
+	device, err := mach.CryptoStore.GetDevice(userID, deviceID)
+	if err != nil {
+		return err
+	} else if device == nil {
+		// fetch if not found
+		usersToDevices := mach.fetchKeys([]id.UserID{userID}, "", true)
+		if devices, ok := usersToDevices[userID]; ok {
+			if device, ok = devices[deviceID]; !ok {
+				return errors.Errorf("Failed to get identity for device %v", deviceID)
+			}
+		} else {
+			return errors.Errorf("Error fetching devices for user %v", userID)
+		}
+	}
+	// create outbound sessions if missing
+	if err := mach.createOutboundSessions(map[id.UserID]map[id.DeviceID]*DeviceIdentity{
+		userID: {
+			deviceID: device,
+		},
+	}); err != nil {
+		return err
+	}
+
+	// get Olm session
+	olmSess, err := mach.CryptoStore.GetLatestSession(device.IdentityKey)
+	if err != nil {
+		return err
+	}
+	if olmSess == nil {
+		return err
```

Comment:
`err` is nil at this point, so it should probably return a different error. (is this case even possible in practice?)

Reply:
It really shouldn't be. Maybe I should just remove the check and let it panic further on if this happens? It would mean something went very wrong.

Reply:
I think this entire `if` was wrong to be there, actually. Removed.